spec/machine/AARCH64/MachineOps.thy

```diff
@@ -280,29 +278,35 @@ definition addressTranslateS1 :: "machine_word \<Rightarrow> machine_word machin
 
 subsection "GIC VCPU Interface"
 
-consts' gic_vcpu_ctrl_hcr_val :: "machine_state \<Rightarrow> 32 word"
 definition get_gic_vcpu_ctrl_hcr :: "32 word machine_monad" where
-  "get_gic_vcpu_ctrl_hcr \<equiv> gets gic_vcpu_ctrl_hcr_val"
+  "get_gic_vcpu_ctrl_hcr \<equiv> gets (\<lambda>ms. vgic_hcr (vcpu_vgic (vcpu_state ms)))"
 
 consts' set_gic_vcpu_ctrl_hcr_impl :: "32 word \<Rightarrow> unit machine_rest_monad"
 definition set_gic_vcpu_ctrl_hcr :: "32 word \<Rightarrow> unit machine_monad" where
-  "set_gic_vcpu_ctrl_hcr w \<equiv> machine_op_lift (set_gic_vcpu_ctrl_hcr_impl w)"
+  "set_gic_vcpu_ctrl_hcr w \<equiv> do
+     modify (vcpu_state_update (vcpu_vgic_update (vgic_hcr_update (\<lambda>_. w))));
+     machine_op_lift (set_gic_vcpu_ctrl_hcr_impl w)
+   od"
 
-consts' gic_vcpu_ctrl_vmcr_val :: "machine_state \<Rightarrow> 32 word"
 definition get_gic_vcpu_ctrl_vmcr :: "32 word machine_monad" where
-  "get_gic_vcpu_ctrl_vmcr \<equiv> gets gic_vcpu_ctrl_vmcr_val"
+  "get_gic_vcpu_ctrl_vmcr \<equiv> gets (\<lambda>ms. vgic_vmcr (vcpu_vgic (vcpu_state ms)))"
 
 consts' set_gic_vcpu_ctrl_vmcr_impl :: "32 word \<Rightarrow> unit machine_rest_monad"
 definition set_gic_vcpu_ctrl_vmcr :: "32 word \<Rightarrow> unit machine_monad" where
-  "set_gic_vcpu_ctrl_vmcr w \<equiv> machine_op_lift (set_gic_vcpu_ctrl_vmcr_impl w)"
+  "set_gic_vcpu_ctrl_vmcr w \<equiv> do
+     modify (vcpu_state_update (vcpu_vgic_update (vgic_vmcr_update (\<lambda>_. w))));
```

Comment:
Probably not needed for this PR but it could be nice to have some wrappers to make it easier to get and modify specific vcpu registers.

Reply:
I did think about this. There was a lemma I wrote (but may have refactored out) for these generic `modify` VCPU updates.

Reply:
It seems mostly the vcpu+vgic combination that could use a wrapper. It's not used *that* often, though, so I'm happy with it either way.